### PR TITLE
Ensure that FQ class names aren't duplicated between jar files in the distribution

### PR DESCRIPTION
**Pull request by tresat**

### CODE REVIEW

File: subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegritySpec.groovy [hunk 1]
```diff
@@ -62,6 +73,15 @@ class DistributionIntegritySpec extends DistributionIntegrationSpec {
...
         jarsWithDuplicateFiles == [:]
+
+        and:
+        def classesDuplicatedBetweenJars = []
+        classesIndex.forEach {className, containingJars ->
+            if (containingJars.size() > 1) {
+                classesDuplicatedBetweenJars << "${className} found in jars: ${containingJars}"
+            }
+        }
+        assert classesDuplicatedBetweenJars.isEmpty()
```
Comment (big-guy) [inline subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegritySpec.groovy:84]: This is a bit simpler
```suggestion
        def duplicateClasses = classesIndex.findAll { it.value.size() > 1 }
        duplicateClasses.isEmpty()
```

File: subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegritySpec.groovy [hunk 2]
```diff
@@ -47,11 +52,17 @@ class DistributionIntegritySpec extends DistributionIntegrationSpec {
 
         when:
         def jarsWithDuplicateFiles = [:]
+        def classesIndex = [:] as HashMap<String, List<String>> // class name -> list of containing jars
         jars.each { jar ->
             new ZipFile(jar).withCloseable {
                 def names = it.entries()*.name
                 def groupedNames = names.groupBy { it }
                 groupedNames.each { name, all ->
+                    if (name.endsWith("class") && !name.endsWith("module-info.class") && !name.endsWith("package-info.class")) {
```
Comment (big-guy) [inline subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegritySpec.groovy:61]: It's a corner case, but this would also accept things named `not_a_class`.

Comment (tresat) [reply subprojects/distributions-integ-tests/src/integTest/groovy/org/gradle/DistributionIntegritySpec.groovy:61]: `.class` would be better.  I'll update.

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Review by big-guy — APPROVED** [review]

**Comment by tresat** [comment]
@bot-gradle test and merge
